### PR TITLE
Update Atmos sources

### PR DESCRIPTION
**Pull request by charleskawczynski**
Updates `AtmosModel` sources, broken off from #610.

Needs review.

### CODE REVIEW

File: src/Atmos/Model/source.jl [hunk 1]
```diff
@@ -38,33 +38,40 @@ struct GeostrophicForcing{FT} <: Source
...
-  FT = eltype(state)
-  z = altitude(m.orientation, aux)
-  coeff = FT(0)
+function atmos_source!(s::RayleighSponge, atmos::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
+  ρ = state.ρ
+  β = FT(0)
   if z >= s.zsponge
-    coeff_top = s.c_sponge * (sinpi(FT(1/2)*(z - s.zsponge)/(s.zmax-s.zsponge)))^FT(4)
-    coeff = min(coeff_top, 1.0)
+      coeff_top = s.c_sponge * (sinpi(FT(1/2)*(z - s.zsponge)/(s.zmax-s.zsponge)))^FT(4)
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]: The design docs use an exponent of 2. Which is correct?

cc: @tapios

File: src/Atmos/Model/source.jl [hunk 2]
```diff
@@ -38,33 +38,40 @@ struct GeostrophicForcing{FT} <: Source
...
+      β = coeff_top
   end
-  source.ρu -= state.ρu * coeff
+
+  u = state.ρu / ρ
+  source.ρu -= ρ * β * (u .- s.u_relaxation)
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]: This could be moved inside the `if` statement

Comment (simonbyrne) [reply src/Atmos/Model/source.jl]:
```suggestion
  source.ρu -=  β_sponge * (state.ρu .- state.ρ .* s.u_relaxation)
```

File: src/Atmos/Model/source.jl (hunk 1 above)
Comment (simonbyrne) [reply src/Atmos/Model/source.jl]:
```suggestion
     β_sponge  = s.α_max * (sinpi(FT(1/2)*(z - s.zsponge)/(s.zmax-s.zsponge)))^FT(4)
```

File: src/Atmos/Model/source.jl [hunk 3]
```diff
@@ -38,33 +38,40 @@ struct GeostrophicForcing{FT} <: Source
...
-function atmos_source!(s::GeostrophicForcing, m::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
-  u = state.ρu / state.ρ
+function atmos_source!(s::GeostrophicForcing, atmos::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
+  ρ = state.ρ
+  u = state.ρu/ρ
   u_geo = SVector(s.u_geostrophic, s.v_geostrophic, 0)
-  source.ρu -= state.ρ * s.f_coriolis * (u - u_geo)
+  ẑ = vertical_unit_vector(atmos.orientation, aux)
+  fkvector = s.f_coriolis * ẑ
+  source.ρu -= ρ * fkvector × (u .- u_geo)
```
Comment (smarras79) [inline src/Atmos/Model/source.jl]: Unless I did the algebra wrong, I believe this should have a `+` sign and not a `-` sign.

The momentum equation should be:

DU/Dt = \rho f (v_g - v)
DV/Dt = \rho f (u   - ug)

but, with the - as written above, the RHS changes.

File: src/Atmos/Model/source.jl [hunk 4]
```diff
@@ -38,33 +38,40 @@ struct GeostrophicForcing{FT} <: Source
...
 Rayleigh Damping (Linear Relaxation) for top wall momentum components
 Assumes laterally periodic boundary conditions for LES flows. Momentum components
 are relaxed to reference values (zero velocities) at the top boundary.
 """
 struct RayleighSponge{FT} <: Source
-  "Domain maximum height [m]"
+  "Domain maximum height `[m]`"
   zmax::FT
-  "Vertical extent at with sponge starts [m]"
+  "Vertical extent at with sponge starts `[m]`"
```
Comment (tapios) [inline src/Atmos/Model/source.jl]:
```suggestion
  "Altitude at with sponge starts `[m]`"
```

File: src/Atmos/Model/source.jl (hunk 1 above)
Comment (tapios) [reply src/Atmos/Model/source.jl]: > The design docs use an exponent of 2. Which is correct?
> 
> cc: @tapios

Either exponent is justifiable (it's just empirical). Klemp & Lilly (1978) use `2` and base their analysis on that, so let's stick with that (as in the design document):
```suggestion
      β_sponge  = s.α_max * (sinpi(FT(1/2)*(z - s.zsponge)/(s.zmax-s.zsponge)))^FT(2)
```

File: src/Atmos/Model/source.jl [hunk 5]
```diff
@@ -38,33 +38,40 @@ struct GeostrophicForcing{FT} <: Source
...
-  RayleighSponge{FT} <: Source
+    RayleighSponge{FT} <: Source
+
 Rayleigh Damping (Linear Relaxation) for top wall momentum components
 Assumes laterally periodic boundary conditions for LES flows. Momentum components
 are relaxed to reference values (zero velocities) at the top boundary.
 """
 struct RayleighSponge{FT} <: Source
-  "Domain maximum height [m]"
+  "Domain maximum height `[m]`"
```
Comment (tapios) [inline src/Atmos/Model/source.jl]:
```suggestion
  "Maximum domain altitude (m)"
```

**Comment by tapios** [comment]
> @simonbyrne I would like to have the ability to decide what sponge to use. As @tapios says, both exponents can work; however, there are a few different damping definitions that may work better in some problems and others for other problems. The 4th exponent case and expression should be the default, but I'd like to give the user the ability to chose other ones. E.g. for deep convection, I will use another type of sponge.

How about making the exponent an argument, with default value 2 (as in the classical Klemp & Lilly paper)?

**Comment by smarras79** [comment]
> > @simonbyrne I would like to have the ability to decide what sponge to use. As @tapios says, both exponents can work; however, there are a few different damping definitions that may work better in some problems and others for other problems. The 4th exponent case and expression should be the default, but I'd like to give the user the ability to chose other ones. E.g. for deep convection, I will use another type of sponge.
> 
> How about making the exponent an argument, with default value 2 (as in the classical Klemp & Lilly paper)?

That is a good option, but it would be nice to also be able to have options for entire sponge expression. It's not really urgent, but having the code ready for that will make it quicker later on.

**Comment by simonbyrne** [comment]
> > 
> 
> How about making the exponent an argument, with default value 2 (as in the classical Klemp & Lilly paper)?

File: src/Atmos/Model/source.jl [hunk 6]
```diff
@@ -38,33 +38,38 @@ struct GeostrophicForcing{FT} <: Source
...
-function atmos_source!(s::GeostrophicForcing, m::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
-  u = state.ρu / state.ρ
+function atmos_source!(s::GeostrophicForcing, atmos::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
+  ρ = state.ρ
+  u = state.ρu/ρ
   u_geo = SVector(s.u_geostrophic, s.v_geostrophic, 0)
-  source.ρu -= state.ρ * s.f_coriolis * (u - u_geo)
+  ẑ = vertical_unit_vector(atmos.orientation, aux)
+  fkvector = s.f_coriolis * ẑ
+  source.ρu -= ρ * fkvector × (u .- u_geo)
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]:
```suggestion
  source.ρu -= fkvector × (state.ρu .- state.ρ .* u_geo)
```

File: src/Atmos/Model/source.jl [hunk 7]
```diff
@@ -38,33 +38,38 @@ struct GeostrophicForcing{FT} <: Source
   u_geostrophic::FT
   v_geostrophic::FT
 end
-function atmos_source!(s::GeostrophicForcing, m::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
-  u = state.ρu / state.ρ
+function atmos_source!(s::GeostrophicForcing, atmos::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
+  ρ = state.ρ
+  u = state.ρu/ρ
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]:
```suggestion
```
Unnecessary with change below

File: src/Atmos/Model/source.jl [hunk 8]
```diff
@@ -38,33 +38,38 @@ struct GeostrophicForcing{FT} <: Source
...
+  u_relaxation::SVector{3,FT}
+  "Sponge exponent"
+  γ::FT
 end
-function atmos_source!(s::RayleighSponge, m::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
-  FT = eltype(state)
-  z = altitude(m.orientation, aux)
-  coeff = FT(0)
+function atmos_source!(s::RayleighSponge, atmos::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
+  β_sponge = FT(0)
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]:
```suggestion
```
no longer needed

File: src/Atmos/Model/source.jl [hunk 9]
```diff
@@ -38,33 +38,38 @@ struct GeostrophicForcing{FT} <: Source
...
   if z >= s.zsponge
-    coeff_top = s.c_sponge * (sinpi(FT(1/2)*(z - s.zsponge)/(s.zmax-s.zsponge)))^FT(4)
-    coeff = min(coeff_top, 1.0)
+    β_sponge = s.α_max * (sinpi(FT(1/2)*(z - s.zsponge)/(s.zmax-s.zsponge)))^FT(γ)
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]:
```suggestion
    r = (z - s.z_sponge)/(s.z_max-s.z_sponge)
    β_sponge = s.α_max * sinpi(r/2)^γ
```

File: src/Atmos/Model/source.jl [hunk 10]
```diff
@@ -38,33 +38,38 @@ struct GeostrophicForcing{FT} <: Source
...
+    RayleighSponge{FT} <: Source
+
 Rayleigh Damping (Linear Relaxation) for top wall momentum components
 Assumes laterally periodic boundary conditions for LES flows. Momentum components
 are relaxed to reference values (zero velocities) at the top boundary.
 """
 struct RayleighSponge{FT} <: Source
-  "Domain maximum height [m]"
+  "Maximum domain altitude (m)"
   zmax::FT
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]:
```suggestion
  z_max::FT
```

File: src/Atmos/Model/source.jl [hunk 11]
```diff
@@ -38,33 +38,38 @@ struct GeostrophicForcing{FT} <: Source
...
-  "Vertical extent at with sponge starts [m]"
+  "Altitude at with sponge starts (m)"
   zsponge::FT
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]:
```suggestion
  z_sponge::FT
```

**Comment by tapios** [comment]
> > > @simonbyrne I would like to have the ability to decide what sponge to use. As @tapios says, both exponents can work; however, there are a few different damping definitions that may work better in some problems and others for other problems. The 4th exponent case and expression should be the default, but I'd like to give the user the ability to chose other ones. E.g. for deep convection, I will use another type of sponge.
> > 
> > 
> > How about making the exponent an argument, with default value 2 (as in the classical Klemp & Lilly paper)?
> 
> That is a good option, but it would be nice to also be able to have options for entire sponge expression. It's not really urgent, but having the code ready for that will make it quicker later on.



> > > @simonbyrne I would like to have the ability to decide what sponge to use. As @tapios says, both exponents can work; however, there are a few different damping definitions that may work better in some problems and others for other problems. The 4th exponent case and expression should be the default, but I'd like to give the user the ability to chose other ones. E.g. for deep convection, I will use another type of sponge.
> > 
> > 
> > How about making the exponent an argument, with default value 2 (as in the classical Klemp & Lilly paper)?
> 
> That is a good option, but it would be nice to also be able to have options for entire sponge expression. It's not really urgent, but having the code ready for that will make it quicker later on.

That would be a good thing to do later. I suspect this will become its own `sponge` module sometime soon.

File: src/Atmos/Model/source.jl (hunk 6 above)
Comment (smarras79) [reply src/Atmos/Model/source.jl]: Please, confirm the sign based on the comment below:

https://github.com/climate-machine/CLIMA/pull/619#discussion_r366054941

File: src/Atmos/Model/source.jl [hunk 12]
```diff
@@ -38,33 +38,36 @@ struct GeostrophicForcing{FT} <: Source
...
 end
-function atmos_source!(s::RayleighSponge, m::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
-  FT = eltype(state)
-  z = altitude(m.orientation, aux)
-  coeff = FT(0)
-  if z >= s.zsponge
-    coeff_top = s.c_sponge * (sinpi(FT(1/2)*(z - s.zsponge)/(s.zmax-s.zsponge)))^FT(4)
-    coeff = min(coeff_top, 1.0)
+function atmos_source!(s::RayleighSponge, atmos::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
+  if z >= s.z_sponge
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]:
```suggestion
  z = altitude(m.orientation, aux)
  if z >= s.z_sponge
```

File: src/Atmos/Model/source.jl [hunk 13]
```diff
@@ -38,33 +38,36 @@ struct GeostrophicForcing{FT} <: Source
...
   v_geostrophic::FT
 end
-function atmos_source!(s::GeostrophicForcing, m::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
-  u = state.ρu / state.ρ
+function atmos_source!(s::GeostrophicForcing, atmos::AtmosModel, source::Vars, state::Vars, aux::Vars, t::Real)
   u_geo = SVector(s.u_geostrophic, s.v_geostrophic, 0)
-  source.ρu -= state.ρ * s.f_coriolis * (u - u_geo)
+  ẑ = vertical_unit_vector(atmos.orientation, aux)
+  fkvector = s.f_coriolis * ẑ
+  source.ρu -= fkvector × (state.ρu .- state.ρ*u_geo)
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]: Need to confirm @smarras79's question on sign here.

File: src/Atmos/Model/source.jl [hunk 14]
```diff
@@ -38,33 +38,37 @@ struct GeostrophicForcing{FT} <: Source
...
-  "Sponge Strength 0 ⩽ c_sponge ⩽ 1"
-  c_sponge::FT
+  "Maximum domain altitude (m)"
+  z_max::FT
+  "Altitude at with sponge starts (m)"
+  z_sponge::FT
+  "Sponge Strength 0 ⩽ α_max ⩽ 1"
+  α_max::FT
+  "Relaxation velocity components"
+  u_relaxation::SVector{3,FT}
```
Comment (simonbyrne) [inline src/Atmos/Model/source.jl]: Is `u_relaxation` global everywhere?

File: src/Atmos/Model/source.jl (hunk 3 above)
Comment (tapios) [reply src/Atmos/Model/source.jl]: Here are the momentum equations: 
Screen Shot 2020-01-13 at 5 12 27 PM
(I.e., opposite sign of what you have above). The Coriolis term above looks correct to me.

Comment (tapios) [reply src/Atmos/Model/source.jl]: Or, another way of looking at it is that the Coriolis force on the RHS is `-2\Omega \times \rho u`, which is what we have (minus the subtraction of the background state).

File: src/Atmos/Model/source.jl (hunk 14 above)
Comment (charleskawczynski) [reply src/Atmos/Model/source.jl]: Well, it's only used in the sponge layer, which seems fine in a box but it's a good question for the sphere.

File: src/Atmos/Model/source.jl (hunk 13 above)
Comment (charleskawczynski) [reply src/Atmos/Model/source.jl]: The sign is [correct](https://github.com/climate-machine/CLIMA/pull/619#discussion_r366111032).

File: src/Atmos/Model/source.jl (hunk 14 above)
Comment (tapios) [reply src/Atmos/Model/source.jl]: On the sphere, we'd want this `u_relaxation` to be the zonal mean (i.e., mean over longitudes)--however this is best implemented.

Comment (simonbyrne) [reply src/Atmos/Model/source.jl]: okay, we may need some extra object to handle that. we can add that later.

**Comment by simonbyrne** [comment]
Need to update all calls to `RayleighSponge`:

https://dev.azure.com/spjbyrne/CLIMA/_build/results?buildId=2105&view=logs&j=4f922444-fdfe-5dcf-b824-02f86439ef14&t=6c1ffda0-4c58-5358-a74f-05db16f3afea&l=468

**Comment by charleskawczynski** [comment]
Yeah, I bolloxed the order

**Pull request merged**